charms/metallb-controller/src/charm.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+"""Controller component for the MetalLB bundle."""
+
+import logging
+import os
+
+from ops.charm import CharmBase
+from ops.framework import StoredState
+from ops.main import main
+from ops.model import (
+    ActiveStatus,
+    BlockedStatus,
+    MaintenanceStatus,
+)
+
+import utils
+
+logger = logging.getLogger(__name__)
+
+
+class MetallbControllerCharm(CharmBase):
+    """MetalLB Controller Charm."""
+
+    _stored = StoredState()
+    # NAMESPACE = os.environ.get("JUJU_MODEL_NAME", 'metallb-system')
+
+    def __init__(self, *args):
+        """Charm initialization for events observation."""
+        super().__init__(*args)
+        self.framework.observe(self.on.start, self.on_start)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.remove, self.on_remove)
+        # -- initialize states --
+        self._stored.set_default(started=False)
+        self._stored.set_default(configured=False)
+        # -- base values --
+        self._stored.set_default(namespace=os.environ["JUJU_MODEL_NAME"])
+        self._stored.set_default(container_image='metallb/controller:v0.9.3')
```

Comment:
If we are going to manage the image ourselves, wouldn't it just work to push it to ROCKS instead of dealing with the image directly in the repo? Unless there's no offline option for ROCKS neither..

Reply:
Some deployments use completely private repos, and I think Juju has some options for working with that when using image resources.  Otherwise, your charm would probably need to have some config to accept the private repository address, and having to manage that across many K8s charms which might all do things slightly differently would get messy.

Reply:
Done with latest commit